File: internal/bot/health.go
package bot

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/cchuter/telegram-trader/internal/blockchain"
	"github.com/cchuter/telegram-trader/internal/galachain"
	"github.com/cchuter/telegram-trader/internal/storage"
)

// HealthStatus represents the overall health of the service
type HealthStatus string

const (
	StatusHealthy   HealthStatus = "healthy"
	StatusDegraded  HealthStatus = "degraded"
	StatusUnhealthy HealthStatus = "unhealthy"
)

// DependencyStatus represents the status of a single dependency
type DependencyStatus struct {
	Name    string `json:"name"`
	Status  string `json:"status"`
	Message string `json:"message,omitempty"`
	Latency string `json:"latency,omitempty"`
}

// HealthResponse represents the health check response
type HealthResponse struct {
	Status       HealthStatus       `json:"status"`
	Timestamp    string             `json:"timestamp"`
	Dependencies []DependencyStatus `json:"dependencies"`
	Version      string             `json:"version"`
	Uptime       string             `json:"uptime"`
}

// HealthChecker provides health check functionality
type HealthChecker struct {
	db         storage.Database
	tonClient  blockchain.Client
	galaClient *galachain.Client
	startTime  time.Time
}

// NewHealthChecker creates a new health checker
func NewHealthChecker(db storage.Database, tonClient blockchain.Client, galaClient *galachain.Client) *HealthChecker {
	return &HealthChecker{
		db:         db,
		tonClient:  tonClient,
		galaClient: galaClient,
		startTime:  time.Now(),
	}
}

// Check performs health checks on all dependencies
func (h *HealthChecker) Check(ctx context.Context) *HealthResponse {
	dependencies := []DependencyStatus{}

	// Check database
	dbStatus := h.checkDatabase(ctx)
	dependencies = append(dependencies, dbStatus)

	// Check TON client
	tonStatus := h.checkTonClient(ctx)
	dependencies = append(dependencies, tonStatus)

	// Check GalaChain gRPC connection
	galaStatus := h.checkGalaChain(ctx)
	dependencies = append(dependencies, galaStatus)

	// Determine overall status
	overallStatus := h.calculateOverallStatus(dependencies)

	// Calculate uptime
	uptime := time.Since(h.startTime).Round(time.Second).String()

	return &HealthResponse{
		Status:       overallStatus,
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
		Dependencies: dependencies,
		Version:      "1.0.0",
		Uptime:       uptime,
	}
}

// checkDatabase checks database connectivity
func (h *HealthChecker) checkDatabase(ctx context.Context) DependencyStatus {
	start := time.Now()

	// Try to query a user session (this will test database connectivity)
	// Using userID 0 which likely doesn't exist, but will test the query
	_, err := h.db.GetUserSession(ctx, 0)

	latency := time.Since(start)

	// "not found" errors are expected and indicate the database is working
	if err != nil {
		errMsg := err.Error()
		if errMsg != "sql: no rows in result set" &&
			errMsg != "user session not found" &&
			errMsg != "user session not found for user_id 0" {
			return DependencyStatus{
				Name:    "database",
				Status:  "unhealthy",
				Message: fmt.Sprintf("Database connection failed: %v", err),
				Latency: latency.String(),
			}
		}
	}

	return DependencyStatus{
		Name:    "database",
		Status:  "healthy",
		Message: "Connected",
		Latency: latency.String(),
	}
}

// checkTonClient checks TON blockchain connectivity
func (h *HealthChecker) checkTonClient(ctx context.Context) DependencyStatus {
	start := time.Now()

	// Try to get balance for a well-known address (TON Foundation)
	// This tests the connection to the TON network
	testAddr := "EQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAM9c" // TON native address

	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()

	_, err := h.tonClient.GetBalance(ctx, testAddr)

	latency := time.Since(start)

	if err != nil {
		return DependencyStatus{
			Name:    "ton_client",
			Status:  "unhealthy",
			Message: fmt.Sprintf("TON client failed: %v", err),
			Latency: latency.String(),
		}
	}

	return DependencyStatus{
		Name:    "ton_client",
		Status:  "healthy",
		Message: "Connected to TON network",
		Latency: latency.String(),
	}
}

// checkGalaChain checks GalaChain gRPC connectivity
func (h *HealthChecker) checkGalaChain(ctx context.Context) DependencyStatus {
	if h.galaClient == nil {
		return DependencyStatus{
			Name:    "galachain_grpc",
			Status:  "unhealthy",
			Message: "GalaChain client not initialized",
		}
	}

	start := time.Now()

	// Try to call GetBalance (even with invalid user ID, this tests connectivity)
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()

	_, err := h.galaClient.GetBalance(ctx, 0)

	latency := time.Since(start)

	if err != nil {
		return DependencyStatus{
			Name:    "galachain_grpc",
			Status:  "unhealthy",
			Message: fmt.Sprintf("gRPC connection failed: %v", err),
			Latency: latency.String(),
		}
	}

	return DependencyStatus{
		Name:    "galachain_grpc",
		Status:  "healthy",
		Message: "Connected to GalaChain service",
		Latency: latency.String(),
	}
}

// calculateOverallStatus determines the overall health based on dependencies
func (h *HealthChecker) calculateOverallStatus(dependencies []DependencyStatus) HealthStatus {
	unhealthyCount := 0
	degradedCount := 0

	for _, dep := range dependencies {
		if dep.Status == "unhealthy" {
			unhealthyCount++
		} else if dep.Status == "degraded" {
			degradedCount++
		}
	}

	// If multiple dependencies are unhealthy, overall is unhealthy
	if unhealthyCount >= 2 {
		return StatusUnhealthy
	}

	// If one dependency is unhealthy or any are degraded, overall is degraded
	if unhealthyCount > 0 || degradedCount > 0 {
		return StatusDegraded
	}

	// All dependencies are healthy
	return StatusHealthy
}

// HTTPHandler returns an HTTP handler for the health check endpoint
func (h *HealthChecker) HTTPHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()

		// Perform health check
		health := h.Check(ctx)

		// Set status code based on health
		statusCode := http.StatusOK
		if health.Status == StatusDegraded {
			statusCode = http.StatusOK // Still 200, but degraded
		} else if health.Status == StatusUnhealthy {
			statusCode = http.StatusServiceUnavailable
		}

		// Write JSON response
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(statusCode)

		if err := json.NewEncoder(w).Encode(health); err != nil {
			http.Error(w, fmt.Sprintf("Failed to encode response: %v", err), http.StatusInternalServerError)
		}
	}
}

// StartHealthServer starts an HTTP server for health checks
func (h *HealthChecker) StartHealthServer(port string) error {
	http.HandleFunc("/health", h.HTTPHandler())
	return http.ListenAndServe(":"+port, nil)
}

File: galachain-service/src/server/health.ts
import * as http from 'http';
import { GSwapClient } from '../gswap/client';

/**
 * Health status types
 */
export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy';

/**
 * Dependency status
 */
export interface DependencyStatus {
  name: string;
  status: string;
  message?: string;
  latency?: string;
}

/**
 * Health check response
 */
export interface HealthCheckResponse {
  status: HealthStatus;
  timestamp: string;
  dependencies: DependencyStatus[];
  version: string;
  uptime: string;
}

/**
 * HealthChecker provides health check functionality for GalaChain service
 */
export class HealthChecker {
  private gswapClient: GSwapClient | null;
  private gswapApiUrl: string;
  private startTime: Date;

  constructor(gswapApiUrl: string, gswapClient?: GSwapClient) {
    this.gswapApiUrl = gswapApiUrl;
    this.gswapClient = gswapClient || null;
    this.startTime = new Date();
  }

  /**
   * Perform health checks on all dependencies
   */
  public async check(): Promise<HealthCheckResponse> {
    const dependencies: DependencyStatus[] = [];

    // Check gswap API connectivity
    const gswapStatus = await this.checkGswapAPI();
    dependencies.push(gswapStatus);

    // Check database connection (placeholder - not implemented in POC)
    const dbStatus = this.checkDatabase();
    dependencies.push(dbStatus);

    // Determine overall status
    const overallStatus = this.calculateOverallStatus(dependencies);

    // Calculate uptime
    const uptimeMs = Date.now() - this.startTime.getTime();
    const uptimeSeconds = Math.floor(uptimeMs / 1000);
    const uptime = this.formatUptime(uptimeSeconds);

    return {
      status: overallStatus,
      timestamp: new Date().toISOString(),
      dependencies,
      version: '1.0.0',
      uptime,
    };
  }

  /**
   * Check gswap API connectivity
   */
  private async checkGswapAPI(): Promise<DependencyStatus> {
    const start = Date.now();

    try {
      // Try to fetch a simple price to test API connectivity
      // Use a well-known pair: GTON/GALA
      if (this.gswapClient) {
        await this.gswapClient.getPrice('GTON', 'GALA');
      } else {
        // If no client provided, just check if the URL is reachable
        const response = await fetch(this.gswapApiUrl, {
          method: 'HEAD',
          signal: AbortSignal.timeout(5000),
        });

        if (!response.ok && response.status !== 404) {
          throw new Error(`API returned status ${response.status}`);
        }
      }

      const latency = `${Date.now() - start}ms`;

      return {
        name: 'gswap_api',
        status: 'healthy',
        message: 'Connected to GSwap API',
        latency,
      };
    } catch (error) {
      const latency = `${Date.now() - start}ms`;

      return {
        name: 'gswap_api',
        status: 'unhealthy',
        message: `GSwap API connection failed: ${error instanceof Error ? error.message : 'Unknown error'}`,
        latency,
      };
    }
  }

  /**
   * Check database connection
   * Placeholder - not implemented in POC
   */
  private checkDatabase(): DependencyStatus {
    // In production, this would check actual database connectivity
    // For POC, we don't have a database in the GalaChain service yet
    return {
      name: 'database',
      status: 'healthy',
      message: 'No database configured (POC)',
    };
  }

  /**
   * Calculate overall health status based on dependencies
   */
  private calculateOverallStatus(dependencies: DependencyStatus[]): HealthStatus {
    let unhealthyCount = 0;
    let degradedCount = 0;

    for (const dep of dependencies) {
      if (dep.status === 'unhealthy') {
        unhealthyCount++;
      } else if (dep.status === 'degraded') {
        degradedCount++;
      }
    }

    // If multiple dependencies are unhealthy, overall is unhealthy
    if (unhealthyCount >= 2) {
      return 'unhealthy';
    }

    // If one dependency is unhealthy or any are degraded, overall is degraded
    if (unhealthyCount > 0 || degradedCount > 0) {
      return 'degraded';
    }

    // All dependencies are healthy
    return 'healthy';
  }

  /**
   * Format uptime in human-readable format
   */
  private formatUptime(seconds: number): string {
    const days = Math.floor(seconds / 86400);
    const hours = Math.floor((seconds % 86400) / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const secs = seconds % 60;

    const parts: string[] = [];
    if (days > 0) parts.push(`${days}d`);
    if (hours > 0) parts.push(`${hours}h`);
    if (minutes > 0) parts.push(`${minutes}m`);
    parts.push(`${secs}s`);

    return parts.join(' ');
  }

  /**
   * HTTP handler for health check endpoint
   */
  public getHTTPHandler(): (_req: http.IncomingMessage, res: http.ServerResponse) => Promise<void> {
    return async (_req: http.IncomingMessage, res: http.ServerResponse) => {
      try {
        // Perform health check
        const health = await this.check();

        // Set status code based on health
        let statusCode = 200;
        if (health.status === 'degraded') {
          statusCode = 200; // Still 200, but degraded
        } else if (health.status === 'unhealthy') {
          statusCode = 503; // Service Unavailable
        }

        // Write JSON response
        res.writeHead(statusCode, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(health, null, 2));
      } catch (error) {
        console.error('Health check failed:', error);
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(
          JSON.stringify({
            status: 'unhealthy',
            error: error instanceof Error ? error.message : 'Unknown error',
          })
        );
      }
    };
  }

  /**
   * Start HTTP server for health checks
   */
  public async startHealthServer(port: number = 8081): Promise<http.Server> {
    const server = http.createServer(this.getHTTPHandler());

    return new Promise((resolve, reject) => {
      server.listen(port, () => {
        console.log(`Health check server listening on port ${port}`);
        resolve(server);
      });

      server.on('error', reject);
    });
  }
}
